python api: allow control over the server certificate verification

shane-kearns opened the pull request:
The --insecure option ("insecure=true" in .zuliprc) disables
verification entirely, similar to other tools like curl.

The --cert_bundle ("cert_bundle=<file>" in .zuliprc) allows
a file to be specified containing the CA certificates to verify
against.
When using self-signed certificates, the server's public key
can be used as the only cerificate in the file.

smarx commented:
*Automated message from Dropbox CLA bot*

@shane-kearns, thanks for the pull request! It looks like you haven't yet signed the Dropbox CLA. Please [sign it here](https://opensource.dropbox.com/cla/) and update the thread so we can consider merging your code.

shane-kearns commented:
This patch is for https://github.com/zulip/zulip/issues/104

timabbott commented on api/zulip/__init__.py:
```diff
@@ -156,7 +171,19 @@ def __init__(self, email=None, api_key=None, config_file=None,
+                insecure_s = config.get("api", "insecure").lower()
+                if insecure_s == "true":
+                    insecure = True
+                elif insecure_s == "false":
```
Can you change `insecure_s` to `insecure_setting` or something similar to match the style in the rest of the Zulip codebase?

timabbott commented on api/zulip/__init__.py:
```diff
@@ -156,7 +171,19 @@ def __init__(self, email=None, api_key=None, config_file=None,
+                    insecure = False
+                else:
+                    raise RuntimeError("insecure is set to '%s', it must be 'true' or 'false' if it is used"
+                                       % (insecure_s,))
```
Maybe it's worth mentioning .zuliprc in this error message for clarity?

timabbott commented:
This looks pretty reasonable; I posted a few small comments on the code.

I think this probably also deserves some updates to the documentation in /integrations and/or /api and a small note in changelog.md.

@shane-kearns if you put "Fixes: #104." in the commit message that will auto-close the ticket when your commit message is merged.

shane-kearns commented:
@timabbott I've updated with your comments, and wrote some documentation in the API readme.
I also spotted the .zuliprc was ignored completely when both --api-key and --user are specified so I fixed that as well.
(The code already prefers command line parameters to config file, but this should save some unexpected traffic on api.zulip.com)

timabbott commented on api/README.md:
```diff
@@ -31,11 +31,23 @@ file is as follows:
     email=<your email address>
     site=<your Zulip server's URI>
+    insecure=<true or false, true means do not verify the server certificate>
+    cert_bundle=<a file containing CA or server certificates to trust>
```
s/a file/path to a file/ ?

timabbott commented on api/README.md:
```diff
@@ -31,11 +31,23 @@ file is as follows:
+which share a home directory.
+
+The command line equivalents for other configuration options are:
+    --site=<your Zulip server's URI>
```
Can we add an example of the format for your Zulip server's URI here, for clarity?  This used to be a corner case and now is the common case and we should make it super clear ;)

timabbott commented on api/README.md:
```diff
@@ -101,3 +113,24 @@ Alternatively, if you don't want to use your ~/.zuliprc file:
+    server.
+    Add "cert_bundle=<filename>" to your .zuliprc file.
+    This option trusts the server / CA certificate only for interaction with
+    the zulip site, and is relatively secure.
```
Two small notes:

* We should probably mention the command-line options here as well
* The formatting for this section looks a bit funny; maybe it should use a bulleted or numbered list formatting?

shane-kearns replied:
I'm not really happy with it either, but markdown does not allow quoted blocks within a list.
I think the version I crossposted with this comment is better formatted.

timabbott commented:
Good catch, and thanks for documenting!  This generally looks great; I posted a few nits and would be happy to merge once those are resolved.

timabbott commented:
Awesome, thanks for the contribution!  Merged.
